module/zfs/zfs_vnops.c

```diff
@@ -2039,13 +2054,22 @@ zfs_mkdir(struct inode *dip, char *dirname, vattr_t *vap, struct inode **ipp,
 	 */
 	zfs_mknode(dzp, vap, tx, cr, 0, &zp, &acl_ids);
 
-	if (fuid_dirtied)
-		zfs_fuid_sync(zfsvfs, tx);
-
 	/*
 	 * Now put new name in parent dir.
 	 */
-	(void) zfs_link_create(dl, zp, tx, ZNEW);
+	error = zfs_link_create(dl, zp, tx, ZNEW);
+	if (error != 0) {
+		zfs_znode_delete(zp, tx);
+		zfs_dirent_unlock(dl);
+		zfs_acl_ids_free(&acl_ids);
+		dmu_tx_commit(tx);
+		iput(ZTOI(zp));
+		ZFS_EXIT(zfsvfs);
+		return (error);
```

Comment:
A similar thing could be done here with an `out` label, it might make the code easier to read.